### PR TITLE
Add test for std::optional, bump tests to C++17

### PR DESCRIPTION
**Pull request by muffgaga**
Should we add a configuration switch for C++17 things like `std::optional`?

### CODE REVIEW

**Comment by muffgaga** [comment]
@obreitwi

**Comment by muffgaga** [comment]
@phispilger You requested this test...

File: tests/optional_parameters_test.py [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import pytest
+import pyoptional_parameters as m
+
+def test_optional_parameters():
+    print dir(m)
+    assert hasattr(m, 'foo')
+    assert m.foo(0) == 0
+    assert m.foo(10) == 10
+    assert m.foo(None) == None
+    assert m.foo() == None
```
Comment (kljohann) [inline tests/optional_parameters_test.py]:
```suggestion
    assert m.foo() is None
```

File: tests/optional_parameters_test.py [hunk 2]
```diff
@@ -0,0 +1,10 @@
+import pytest
+import pyoptional_parameters as m
+
+def test_optional_parameters():
+    print dir(m)
+    assert hasattr(m, 'foo')
+    assert m.foo(0) == 0
+    assert m.foo(10) == 10
+    assert m.foo(None) == None
```
Comment (kljohann) [inline tests/optional_parameters_test.py]: Please use `is None`.
```suggestion
    assert m.foo(None) is None
```

File: tests/optional_parameters_test.py [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import pytest
+import pyoptional_parameters as m
+
+def test_optional_parameters():
+    print dir(m)
```
Comment (kljohann) [inline tests/optional_parameters_test.py]: Please remove this print statement.
```suggestion
```

File: tests/optional_parameters.cpp [hunk 4]
```diff
@@ -0,0 +1,6 @@
+#include "optional_parameters.h"
+
+std::optional<int> GENPYBIND(visible) foo(std::optional<int> o)
+{
+	return o;
```
Comment (kljohann) [inline tests/optional_parameters.cpp]: Please clang-format this file using the configuration contained in the repo.

**Review by kljohann — CHANGES_REQUESTED** [review]
Thanks for the PR! Some minor comments. Can you make it more clear that the C++17 flags only apply to the tests? For example "Add test for std::optional, bump tests to C++17"?
Can you also change `std=c++14` to `std=c++17` in `tests/genpybind_waf.py`?

**Comment by muffgaga** [comment]
@kljohann ACK, will handle → I'll also test-drive on clang7.

**Comment by muffgaga** [comment]
> @kljohann ACK, will handle → I'll also test-drive on clang7.

clang7 works for me -> @phispilger fixes the other stuff plz

File: tests/optional_parameters.h [hunk 5]
```diff
@@ -0,0 +1,18 @@
+#pragma once
+
+#if !__has_include("optional")
+// clang 5 does not support std::optional :(
```
Comment (kljohann) [inline tests/optional_parameters.h]: I think this is not clang's fault but rather is necessary due to the old version of libstdc++ in the CI image. (It's still using gcc 6.3.0.) Maybe that's also what you ran into during your local testing?

Comment (muffgaga) [reply tests/optional_parameters.h]: ACK, and dropped comment → `--gcc-toolchain=` option solves it.

Comment (muffgaga) [reply tests/optional_parameters.h]: And a correct installation of clang (`-DGCC_PREFIX_PATH`) also fixes the build without providing `--gcc-toolchain`.

**Review by kljohann — APPROVED** [review]

**Comment by kljohann** [comment]
Merged, thanks!

**Pull request closed without merging**